// react-router-dom/es/Link.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from "react";
import PropTypes from "prop-types";
import { createLocation } from "history";
import invariant from "invariant";

import RouterContext from "./RouterContext";

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

/**
 * The public API for rendering a history-aware <a>.
 */

var Link = function (_React$Component) {
  _inherits(Link, _React$Component);

  function Link() {
    _classCallCheck(this, Link);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  Link.prototype.handleClick = function handleClick(event, context) {
    if (this.props.onClick) this.props.onClick(event);

    if (!event.defaultPrevented && // onClick prevented default
    event.button === 0 && // ignore everything but left clicks
    !this.props.target && // let browser handle "target=_blank" etc.
    !isModifiedEvent(event) // ignore clicks with modifier keys
    ) {
        event.preventDefault();

        var method = this.props.replace ? context.history.replace : context.history.push;

        method(this.props.to);
      }
  };

  Link.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        innerRef = _props.innerRef,
        replace = _props.replace,
        to = _props.to,
        props = _objectWithoutProperties(_props, ["innerRef", "replace", "to"]); // eslint-disable-line no-unused-vars

    return React.createElement(
      RouterContext.Consumer,
      null,
      function (context) {
        !context ? process.env.NODE_ENV !== "production" ? invariant(false, "You should not use <Link> outside a <Router>") : invariant(false) : void 0;

        var location = typeof to === "string" ? createLocation(to, null, null, context.location) : to;
        var href = location ? context.history.createHref(location) : "";

        return React.createElement("a", _extends({}, props, {
          onClick: function onClick(event) {
            return _this2.handleClick(event, context);
          },
          href: href,
          ref: innerRef
        }));
      }
    );
  };

  return Link;
}(React.Component);

Link.defaultProps = {
  replace: false
};


if (process.env.NODE_ENV !== "production") {
  var toType = PropTypes.oneOfType([PropTypes.string, PropTypes.object]);

  Link.propTypes = {
    innerRef: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    onClick: PropTypes.func,
    replace: PropTypes.bool,
    target: PropTypes.string,
    to: toType.isRequired
  };
}

export default Link;
