// react-router/es/Prompt.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from "react";
import PropTypes from "prop-types";
import invariant from "invariant";

import RouterContext from "./RouterContext";

var Block = function (_React$Component) {
  _inherits(Block, _React$Component);

  function Block(props) {
    _classCallCheck(this, Block);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.release = props.method(props.message);
    return _this;
  }

  Block.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    if (prevProps.message !== this.props.message) {
      this.release();
      this.release = this.props.method(this.props.message);
    }
  };

  Block.prototype.componentWillUnmount = function componentWillUnmount() {
    this.release();
  };

  Block.prototype.render = function render() {
    return null;
  };

  return Block;
}(React.Component);

if (process.env.NODE_ENV !== "production") {
  var messageType = PropTypes.oneOfType([PropTypes.func, PropTypes.string]);

  Block.propTypes = {
    method: PropTypes.func.isRequired,
    message: messageType.isRequired
  };
}

/**
 * The public API for prompting the user before navigating away from a screen.
 */
function Prompt(props) {
  return React.createElement(
    RouterContext.Consumer,
    null,
    function (context) {
      !context ? process.env.NODE_ENV !== "production" ? invariant(false, "You should not use <Prompt> outside a <Router>") : invariant(false) : void 0;

      return props.when ? React.createElement(Block, { method: context.history.block, message: props.message }) : null;
    }
  );
}

Prompt.defaultProps = {
  when: true
};

if (process.env.NODE_ENV !== "production") {
  var _messageType = PropTypes.oneOfType([PropTypes.func, PropTypes.string]);

  Prompt.propTypes = {
    when: PropTypes.bool,
    message: _messageType.isRequired
  };
}

export default Prompt;
